qiskit/assembler/assemble_circuits.py

```diff
@@ -145,46 +168,76 @@ def _assemble_pulse_gates(circuit, run_config):
     return QasmExperimentCalibrations(gates=calibrations), pulse_library
 
 
-def _extract_common_calibrations(experiments):
+def _extract_common_calibrations(
+        experiments: List[QasmQobjExperiment]
+) -> Tuple[List[QasmQobjExperiment], Optional[QasmExperimentCalibrations]]:
     """Given a list of ``QasmQobjExperiment``s, each of which may have calibrations in their
     ``config``, collect common calibrations into a global ``QasmExperimentCalibrations``
     and delete them from their local experiments.
+
+    Args:
+        experiments: The list of Qasm experiments that are being assembled into one qobj
+
+    Returns:
+        The input experiments with modified calibrations, and common calibrations, if there
+        are any
     """
+    def index_calibrations() -> Dict[int, List[Tuple[int, GateCalibration]]]:
+        """Map each calibration to all experiments that contain it."""
+        exp_indices = defaultdict(list)
+        for exp_idx, exp in enumerate(experiments):
+            for gate_cal in exp.config.calibrations.gates:
+                # They must be keyed on the hash or identical cals will be indexed separately
+                exp_indices[hash(gate_cal)].append((exp_idx, gate_cal))
+        return exp_indices
+
+    def collect_common_calibrations() -> List[GateCalibration]:
+        """If a gate calibration appears in all experiments, collect it."""
+        common_calibrations = []
+        for _, exps_w_cal in exp_indices.items():
+            if len(exps_w_cal) == len(experiments):
+                _, gate_cal = exps_w_cal[0]
+                common_calibrations.append(gate_cal)
+        return common_calibrations
+
+    def remove_common_gate_calibrations(exps: List[QasmQobjExperiment]) -> None:
+        """For calibrations that appear in all experiments, remove them from the individual
+        experiment's ``config.calibrations``."""
+        for _, exps_w_cal in exp_indices.items():
+            if len(exps_w_cal) == len(exps):
+                for exp_idx, gate_cal in exps_w_cal:
+                    exps[exp_idx].config.calibrations.gates.remove(gate_cal)
```

Comment:
This seems like a case of the "Law of Demeter" but it may be unavoidable since these classes are just data containers and we do not want methods on them.

Reply:
thats a good point